### DISCUSSION TITLE
BgB vs QLKNN

### DISCUSSION THREAD

**Discussion by theo-brown** [google-deepmind/torax#446] [General]
I'm running the `iterhybrid_predictor_corrector` scenario in `torax/examples`, trying to sense-check the Bohm-gyroBohm model against QLKNN. This is one of the first times I've actually run BgB in TORAX in a 'realistic' scenario.

This is the comparison so far, after manual twiddling of the BgB parameters (QLKNN in dashed):
![image](https://github.com/user-attachments/assets/6de7297c-2a0e-4504-be7e-b0d1ff177d7b)
It generally looks pretty good, although I'm never sure what behaviour of QLKNN is signal vs noise so don't have 100% trust in my twiddled BgB params. If I was doing this properly I suppose I could try and dig out an ITER BgB simulation...

The only glaring defects are:
1. Missing bootstrap current around `rho=0.2` in the BgB model.
2. Electron density too low in the BgB model.

I'll let someone with more physics knowledge than me tackle (1), as to my knowledge it's not a direct consequence of the BgB vs QLKNN model per se, but may be instead to do with the particular shapes (esp. gradients) in the profiles. 

(2) is the question I really want to answer at the moment, because the particle diffusivity and convectivity are properties of the transport model.

Looking at my plot, the density in the core is too low compared to QLKNN, which suggests that my coefficients are too low (the source is off axis, so low coeffs = slower diffusion into the core).

Looking at the QLKNN run vs my run, the `d_face_el` terms are both at their clipped values, but the `v_face_el` are noticeably different.

In QLKNN, it's negative:
![image](https://github.com/user-attachments/assets/d1232fd0-db44-43a0-8f06-4fd5174bdeae)

In BgB, it's positive:
![image](https://github.com/user-attachments/assets/a4a73f62-3c62-4ec8-a700-259b917765b5)

**My suspicion is that this is a simple sign error**. As I was using JETTO as my template for the BgB implementation, I think that the convectivity I implemented is a pinch velocity - am I right in thinking that this means it is *inwards*? Does TORAX treat `v_face_el` as a pinch term, or is it a general convectivity (ie could be inwards or outwards)?

**Comment by jcitrin** [comment]
> although I'm never sure what behaviour of QLKNN is signal vs noise so don't have 100% trust in my twiddled BgB params. 

There is indeed a bit more QLKNN jitter with the predictor-corrector. You can try the following to smooth things:

(i) add some Gaussian smoothing to the transport coefficients, e.g. add `'smoothing_sigma': 0.1,` to the transport dict.
(iia) use the nonlinear solver (tests/test_data/test_iterhybrid_newton.py) which will be smoother
(iib) alternatively increase the chi_per and d_per values in the stepper dict (pereverzev-corrigan terms) which will increase smoothness of the predictor_corrector run at the cost of potentially artificially increasing the time until stationary state is reached. 

Furthermore, you can consider:

1. turning off current diffusion to remove the longer current diffusion timescale and also freezing q and s for both models
2. extending the simulations times a bit more to ensure that you reach a stationary state for temperature and density

Regarding your point (1), I indeed wouldn't be concerned about the bootstrap current differences around rho~0.2. It's very sensitive to the exact temperature and density gradients, and this zone is where the transport model is transitioning to from QLKNN/BgB predicitons to the "inner_patch" region. QLKNN tends to predict no/low transport inside rho<0.3 or so, leading to the Ti sharp gradient in the narrow zone until the inner_patch is reached, which I think is mostly responsible for the j_bs bump.

Finally, about point (2). Yes, TORAX V is just convection, so negative V is inward convection ("pinch") and positive V is outward convection. So if in JETTO you have a "pinch term" which is defined positive but actually leads to inward transport, then in TORAX this just needs to be a negative V.

**Reply by theo-brown** [reply]
> QLKNN tends to predict no/low transport inside rho<0.3 or so, leading to the Ti sharp gradient in the narrow zone until the inner_patch is reached, which I think is mostly responsible for the j_bs bump.

Yes, this is exactly what I thought was going on!

> Finally, about point (2). Yes, TORAX V is just convection, so negative V is inward convection ("pinch") and positive V is outward convection. So if in JETTO you have a "pinch term" which is defined positive but actually leads to inward transport, then in TORAX this just needs to be a negative V.

PR to fix the sign in #447.

**Comment by theo-brown** [comment]
I've rerun with a bunch of your suggestions (transport model smoothing, nonlinear solver, longer timescales) and plotted with the new plot tools (very nice!). 
The results haven't changed that much (see below, QLKNN in dashed, BgB in solid)

![Figure_1](https://github.com/user-attachments/assets/75a2a625-53e2-4849-9116-fea3c8e79f36)

q-profile for QLKNN is a bit dodgy in the core this time around, but that's probably still due to the inner patch.
Discrepancy in Te, Ti, and ne between BgB and QLKNN is still ~10-15, leading to significantly lower fusion power in the BgB case. 

The `v_face_el` discrepancy can now be seen in these plots. I'm going to dig around in the JETTO code to find out whether there's a unit conversion I missed.

**Comment by jcitrin** [comment]
Yeah a current hole is developing in the QLKNN simulation. Probably exasperated by the much higher bootstrap current which is I think also due to the density peaking difference.

You can try reducing somewhat fext (external current fraction). That should help with the current hole.

Definitely worth checking on the magnitude of BgB pinch term. Note that the pure transport-induced density peaking factor in BgB is not configurable. e.g., in a source-free system R/Lne ~ -RV/D , and the V/D ratio is fixed in the BgB model. So, while it's worth checking the magnitude of V, if we are confident on the ultimate value of V, then what you get for the peaking is what you get.

In the end these are very different transport models and even with some tuning there will be a gap.

**Comment by theo-brown** [comment]
I have the following comparison:

| Code   | Tokamak | Transport model | $v$     |
|--------|---------|-----------------|---------|
| JETTO  | STEP    | BgB             | ~2e-3   |
| TORAX  | ITER    | QLKNN           | ~6e-2   |
| TORAX  | ITER    | BgB             | ~5e-5   |

Looking at the JETTO code, I'm fairly certain that there aren't any extra conversion factors we're missing, and that the description in Tholerus is correct.

**Comment by jcitrin** [comment]
Would be good to compare say -RV/D with BgB for a couple of different machines, JETTO vs TORAX.

I also don't see any issue with the formula. I think we just leave it for now but this really does motivate more stringent benchmarks between JETTO and TORAX, e.g. for the STEP case.

**Reply by theo-brown** [reply]
100% agree! Good to have it flagged as a thing that will need a more detailed look once we can run direct comparisons. Given the sign was definitely wrong, do you want to merge the fix in #447 or leave that PR open as a placeholder for "investigate RV/D"? Or the further investigation could be logged as an issue.

**Reply by jcitrin** [reply]
We can merge and close #447 with the sign fix, and open an issue for investigating BgB (and also the ECCD model for good measure) in closer benchmarks against JETTO
